### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `tl.ThreadSync` silently drops an in-loop `__syncthreads()` inside a `threadIdx`-dependent `if`, racing to a wrong result

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2612]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (source at `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

Reproduced on NVIDIA H100 (sm_90) and NVIDIA L40S (sm_89), CUDA 13.0 / 12.9, tilelang 0.1.11. The defect is in the target-independent `thread_storage_sync` TIR pass, so it is architecture-independent; it reproduces identically on both GPUs (same emitted CUDA, same all-wrong result).

### Problem description

When the `thread_storage_sync` planner decides a `__syncthreads()` is required **inside** a `threadIdx`-dependent `if`, and the condition is either runtime-non-uniform (`if Mask[tx] != 0`) or a strided thread set that isn't a single contiguous warp run (`if tx % 4 == 0`), the pass **erases the in-`if` barrier and moves a single barrier to *before* the `if`**. If the write and its dependent cross-thread read both live inside the guard (e.g. inside a loop in the guarded body), the hoisted barrier no longer sits between them: the shared-memory write→read becomes an unsynchronized data race with a silently wrong, nondeterministic result — no error, no visible warning.

The user writes no barriers; `ThreadSync` owns barrier insertion here. The identical body *without* the guard is synchronized correctly (see the control in the repro), so the only difference between a correct kernel and a silently racing one is wrapping the body in a `threadIdx`-dependent `if`.

<details><summary>Emitted CUDA for the minimal repro — the barrier is before the loop, not inside it</summary>

For `if Mask[tx] != 0` around a loop that writes `S[tx]` then reads the partner `S[127-tx]`, `thread_storage_sync` splits the `if` and emits **one** `__syncthreads()` between the two halves — *before* the loop — and **none inside** the loop:

```cuda
Out[threadIdx.x] = -999.0f;
if (Mask[threadIdx.x] != 0) {
  acc[0] = 0.0f;
}
__syncthreads();                       // hoisted here
if (Mask[threadIdx.x] != 0) {
  for (int i = 0; i < 2; ++i) {
    S[threadIdx.x] = A[i * 128 + threadIdx.x];      // WRITE
    acc[0] = acc[0] + S[127 - threadIdx.x];         // cross-thread READ, no sync between
  }
  Out[threadIdx.x] = acc[0];
}
```

The same body with the guard removed emits **two** in-loop `__syncthreads()` (WAR before the write, RAW after) and is correct:

```cuda
for (int i = 0; i < 2; ++i) {
  __syncthreads();
  S[threadIdx.x] = A[i * 128 + threadIdx.x];
  __syncthreads();
  acc[0] = acc[0] + S[127 - threadIdx.x];
}
```

</details>

<details><summary>Trigger boundary (each cell run this session; sm_90 and sm_89)</summary>

Block of 128 threads, shared `S[128]`, 2-iteration loop `S[tx]=A[i,tx]; acc+=S[127-tx]`, cross-thread read of the partner.

| guard around the body | active set | in-loop `__syncthreads` | result |
|---|---|---|---|
| none (control) | all 128 | 2 | correct (0/128 wrong) |
| `if Mask[tx] != 0`, all-ones mask (runtime, non-uniform) | all 128 | **0** | **wrong on every run** (64–128 threads wrong, varies run-to-run — it is a genuine barrier-less race) |
| `if tx % 4 == 0` (strided, 32 active = multiple of 32) | 32, one per 4 lanes, spanning all 4 warps | **0** | **wrong (max_err 3.3)** |
| `if tx % 2 == 0` (strided, 64 active) | 64 | **0** | **wrong (max_err 3.5)** |
| 2D block, `if ty % 2 == 0` (even rows) | even rows | 2 (in-loop) | correct — hoist not taken |
| 2D block, `if ty < TY/2` (contiguous rows) | first half | 2 (partial, in-loop) | correct — partial-sync lowered |

The `tx % 4 == 0` cell is the sharpest: its active count is 32 (a multiple of 32), so it is **not** the `%32`-count case of #2556; the barrier is dropped purely because the 32 active threads are not a contiguous warp run, so `AnalyzeCondition` cannot count `warp_size` consecutive satisfying lanes and sets `requires_hoist`.

Under `compute-sanitizer --tool=racecheck --racecheck-report=all`, the runtime-mask kernel reports RAW hazards on `S` at the cross-thread read line, and the strided kernel reports WAR hazards — the missing-sync confirmed independently of whether the value happened to come out wrong on a given run.

</details>

### Reproducible example code

```python
import os; os.environ.setdefault("CUDA_HOME", "/tmp/cuda_home_r9")
import torch, tilelang
import tilelang.language as T

THREADS, ITERS = 128, 2
PC = {"tl.device_compile_flags": ["-DCCCL_DISABLE_CTK_COMPATIBILITY_CHECK"]}  # env-specific, not part of the bug

def build(guarded):
    @T.prim_func
    def main(A: T.Tensor((ITERS, THREADS), "float32"),
             Mask: T.Tensor((THREADS,), "int32"),
             Out: T.Tensor((THREADS,), "float32")):
        with T.Kernel(1, threads=THREADS) as bx:
            S = T.alloc_shared((THREADS,), "float32")
            acc = T.alloc_local((1,), "float32")
            tx = T.get_thread_binding(0)
            Out[tx] = -999.0
            if guarded:
                if Mask[tx] != 0:                       # runtime, non-uniform guard
                    acc[0] = 0.0
                    for i in T.serial(ITERS):
                        S[tx] = A[i, tx]                # WRITE
                        acc[0] += S[(THREADS - 1) - tx] # cross-thread READ (partner)
                    Out[tx] = acc[0]
            else:                                       # same body, no guard = control
                acc[0] = 0.0
                for i in T.serial(ITERS):
                    S[tx] = A[i, tx]
                    acc[0] += S[(THREADS - 1) - tx]
                Out[tx] = acc[0]
    return tilelang.compile(main, target="cuda", pass_configs=PC)

torch.manual_seed(0)
A = torch.randn(ITERS, THREADS, device="cuda", dtype=torch.float32)
Mask = torch.ones(THREADS, device="cuda", dtype=torch.int32)   # every thread active
for guarded in (False, True):
    k = build(guarded)
    Out = torch.full((THREADS,), -999.0, device="cuda")
    k(A, Mask, Out); torch.cuda.synchronize()
    nsync = k.get_kernel_source().count("__syncthreads")
    exp = A[:, THREADS - 1].sum().item()                       # value thread 0 should read (partner=127)
    print(f"guarded={guarded} n_syncthreads={nsync} Out[0]={Out[0].item():.4f} exp={exp:.4f}")
# guarded=False -> Out[0] == exp   (control, correct)
# guarded=True  -> result differs from the reference on every run (the specific wrong value
#                  varies run-to-run: barrier-less race). Oracle below = barrier COUNT, not a fixed value.
```

Run under the detector to confirm the hazard directly. NOTE: the barrier is dropped on every run (deterministic), so the result is wrong on every run, but the *specific* wrong value is scheduling-dependent (a real race) — so the robust oracle is the barrier COUNT in the emitted CUDA (0 in-loop syncs vs 2 for the control), not a fixed wrong value:

```
compute-sanitizer --tool=racecheck --racecheck-report=all python repro.py
# ... Potential RAW hazard detected at __shared__ ... (the acc[0] += S[...] line)
```

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong (a compile-time `LOG(WARNING)` "[ThreadSync] Hoisting sync from inside if to before if" is emitted at a log level that is suppressed by default, so nothing reaches the user).
```

### Expected behavior

`ThreadSync` should keep the write→read ordering it identified. When the guard is a runtime-non-uniform or strided condition, hoisting a *single* barrier out of the `if` protects the entry into the guarded region but not a cross-thread dependency *inside* it, so it under-synchronizes. The 2D contiguous-row case in the boundary table shows the pass already has a correct alternative — `__sync_thread_partial` lowered in-place — and the un-guarded control shows the pass emits the correct in-loop barriers when the same body is not wrapped; the guarded case should reach the same in-place synchronization (or, if that is genuinely not lowerable for this condition, decline to compile rather than emit a racing kernel). The current hoist trades a deadlock (which the hoist was added to avoid) for a silent data race.

### Additional context

**Root cause.** In [`ThreadStorageSyncPlanner::VisitStmt_(const IfThenElseNode*)`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/thread_storage_sync.cc#L945-L960), when a sync was inserted inside the `if` and the condition is `(depends_on_runtime && !is_block_uniform) || requires_hoist`, the in-`if` syncs are `erase`d and a single `insert_syncs(op)` is placed before the `if`. `requires_hoist` is set in [`AnalyzeCondition`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/thread_storage_sync.cc#L456-L476) whenever the Z3 prover cannot count `warp_size` **consecutive** satisfying lanes — which a strided set (`tx % 4 == 0`) fails even though its total count is a multiple of 32. In both branches the hoist assumes the erased barrier only guarded region entry; it does not check for a write→read dependency living *inside* the guarded body, which the barrier was actually protecting.

<details><summary>Suggested fix (direction only — verified by the pass's own behavior, not built)</summary>

Before erasing the in-`if` syncs, check whether the guarded body contains a cross-thread shared write→read that those syncs were satisfying. If it does, the single hoisted barrier is insufficient: either keep an in-place barrier lowered for the active set (the `__sync_thread_partial` path that already works for the contiguous-row case), or, when the condition truly cannot be lowered safely, fail the compilation rather than silently emit an under-synchronized kernel. The deadlock the hoist guards against and the race it introduces are the two horns; a silent wrong result should not be the resolution of either.

</details>

**Distinct from #2556.** #2556 is `ProcessSharedSync` returning an empty `Stmt()` for an active-thread **count** that isn't a multiple of 32; it is a different code path and does not fire here (the `tx % 4 == 0` cell has exactly 32 active threads yet still drops the barrier, via this `if`-hoist path). I searched the open and closed tracker and found no existing report of the barrier being dropped by the `if`-condition hoist.

**Provenance.** The hoist-and-erase mechanism was introduced in #1781 ("Fix thread storage sync conflict detection for loop carry write-after-read", merged 2026-02-04); the condition was later generalized to `(depends_on_runtime && !is_block_uniform) || requires_hoist` in #1923 ("Support ptr-table grouped GEMM kernels", merged 2026-03-12). Present since #1781 shipped; the current form dates to #1923. Reproduced on 0.1.11 (`cd37ed5f`) this session on sm_90 and sm_89.
